sdk/identity/azure-identity/azure/identity/credentials.py

```diff
@@ -15,42 +20,115 @@
 
 if TYPE_CHECKING:
     # pylint:disable=unused-import
-    from typing import Any, Dict, Iterable, Mapping, Optional
-    from azure.core.pipeline.policies import HTTPPolicy
+    from typing import Any, Iterable, List, Mapping, Optional, Union
     from azure.core.credentials import SupportsGetToken
 
 # pylint:disable=too-few-public-methods
 
 
-class _ClientCredentialBase(object):
-    _OAUTH_ENDPOINT = "https://login.microsoftonline.com/{}/oauth2/v2.0/token"
+class ClientSecretCredential(ClientSecretCredentialBase):
+    """Authenticates with a client secret"""
 
-    def __init__(self, client_id, tenant_id, config=None, policies=None, **kwargs):
-        # type: (str, str, Optional[Configuration], Optional[Iterable[HTTPPolicy]], Mapping[str, Any]) -> None
-        if not client_id:
-            raise ValueError("client_id")
-        if not tenant_id:
-            raise ValueError("tenant_id")
-        self._client = AuthnClient(self._OAUTH_ENDPOINT.format(tenant_id), config, policies, **kwargs)
-        self._form_data = {}  # type: Dict[str, str]
+    def __init__(self, client_id, secret, tenant_id, config=None, **kwargs):
+        # type: (str, str, str, Optional[Configuration], Mapping[str, Any]) -> None
+        super(ClientSecretCredential, self).__init__(client_id, secret, tenant_id, **kwargs)
+        self._client = AuthnClient(OAUTH_ENDPOINT.format(tenant_id), config, **kwargs)
 
     def get_token(self, scopes):
         # type: (Iterable[str]) -> str
-        data = self._form_data.copy()
-        data["scope"] = " ".join(scopes)
         token = self._client.get_cached_token(scopes)
         if not token:
-            return self._client.request_token(scopes, form_data=data)
+            data = dict(self._form_data, scope=" ".join(scopes))
+            token = self._client.request_token(scopes, form_data=data)
         return token
 
 
-class ClientSecretCredential(_ClientCredentialBase):
-    def __init__(self, client_id, secret, tenant_id, config=None, **kwargs):
-        # type: (str, str, str, Optional[Configuration], Mapping[str, Any]) -> None
-        if not secret:
-            raise ValueError("secret")
-        super(ClientSecretCredential, self).__init__(client_id, tenant_id, config, **kwargs)
-        self._form_data = {"client_id": client_id, "client_secret": secret, "grant_type": "client_credentials"}
+class CertificateCredential(CertificateCredentialBase):
+    """Authenticates with a certificate (thumbprint and PEM-encoded private key)"""
+
+    def __init__(self, client_id, tenant_id, private_key, thumbprint, config=None, **kwargs):
```

Comment:
>def __init__(self, client_id, tenant_id, private_key, thumbprint, config=None, **kwargs): [](start = 4, length = 89)

private key should be the certificate path. Also the path and thumbprint would both be optional (but you have to specify one).  That being said the usage of thumbprint is that I can find the cert with that thumbprint in the cert store, which we're not supporting yet so perhaps just remove thumbprint for now?